tower-http/src/sensitive_header.rs

```diff
@@ -155,78 +177,93 @@ where
     }
 
     fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
-        if let Some(value) = req.headers_mut().get_mut(&self.header) {
-            value.set_sensitive(true);
+        for header in self.headers.clone() {
```

Comment:
I think this clone is unnecessary and could be avoided --- we only need an `&HeaderName` to call `headers_mut().get_mut()`. It would be nice to not clone header names, which may be strings, on every call.

What if we changed the
```rust
where
    I: IntoIterator<Item = HeaderName> + Clone
```
bound to 
```rust
where
    for<'a> &'a I: IntoIterator<Item = &'a HeaderName>
```

That way, we could just write
```rust
for header in &self.headers {
```
here, and avoid having to clone all the headers (and the data structure they're stored in, which might be a `Vec` or similar) in every `call`.

Reply:
I can't quite get this to work. After updating the bounds and applying the layer like so

```rust
.layer(SetSensitiveHeaderLayer::new(once(header::AUTHORIZATION)))
```

I get this error

```
 1  error[E0277]: `&'a std::iter::Once<hyper::header::HeaderName>` is not an iterator
    --> examples/tonic-key-value-store/src/main.rs:139:16
     |
 139 |         .layer(SetSensitiveHeaderLayer::new(once(header::AUTHORIZATION)))
     |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `&'a std::iter::Once<hyper::header::HeaderName>` is not an iterator
     |
    ::: /Users/david.pedersen/dev/major/tower-http/tower-http/src/sensitive_header.rs:69:37
     |
 69  |         for<'a> &'a I: IntoIterator<Item = &'a HeaderName>,
     |                                     --------------------- required by this bound in `tower_http::sensitive_header::SetSensitiveHeaderLayer::<I>::new`
     |
     = help: the trait `std::iter::Iterator` is not implemented for `&'a std::iter::Once<hyper::header::HeaderName>`
     = note: `std::iter::Iterator` is implemented for `&'a mut std::iter::Once<hyper::header::HeaderName>`, but not for `&'a std::iter::Once<hyper::header::HeaderName>`
     = note: required because of the requirements on the impl of `for<'a> std::iter::IntoIterator` for `&'a std::iter::Once<hyper::header::HeaderName>`
```

Not sure if thats fixable. Do you know? I've pushed the code in case you wanna play around.

Also, looking at the `HeaderName` implementation it seems to be backed by either a `Copy` constant or a `ByteStr` which is cheap to clone (thanks to `Bytes`). If the iterator is `vec![...].into_iter()` that isn't cheap but I've been using the new `std::array::IntoIter::new([...])` which should avoid allocations from a `Vec`.

Reply:
> Not sure if thats fixable. Do you know? I've pushed the code in case you wanna play around.

Hmm, I think it's specifically `std::iter::once` that's the issue here. An array with one element ought to work, [as should `Option`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#impl-IntoIterator-1).

> Also, looking at the `HeaderName` implementation it seems to be backed by either a `Copy` constant or a `ByteStr` which is cheap to clone (thanks to `Bytes`). If the iterator is `vec![...].into_iter()` that isn't cheap but I've been using the new `std::array::IntoIter::new([...])` which should avoid allocations from a `Vec`

Yeah, the issue here IMO is not cloning the header names themselves, but the type they're stored in. My concern specifically is that because this function is generic and you can pass in anything, users will easily construct `SetSensitiveHeaderLayer`s with a `Vec<HeaderName>`, which results in the `Vec` itself getting cloned a bunch.

Another option is to change the `IntoIterator` bound to `AsRef<[&HeaderName]>`. Then, we could just call `as_ref()` rather than cloning, and iterate over the slice. This should work fine.

The downside of _that_ approach is that it doesn't allow you to pass in a `HashSet` or something (though I'm not immediately sure why you would do that instead of just using a slice/`Vec`?) and it wouldn't work with `std::iter::once` _or_ with an `Option` --- a single header would _have_ to be a one-element slice...